From 32a9f3ffebb34134cd7e2be54b246f4ea8495c32 Mon Sep 17 00:00:00 2001
From: Zaikuo Wang <zaikuo.wang@intel.com>
Date: Thu, 12 Mar 2015 14:08:00 +0800
Subject: [PATCH 296/307] ov5693: add delay time and retry startup when powerup

On CHT_PR2, ov5693 power on failed when run MTBF;
This issue did not happen at our site, I add some solution
to make powerUp to be more robust:
1. increase delay time between DOVDD and PWDN
2. retry to powerUp at most 5 times if startUp failed at first time

Change-Id: I084b48d8ff6ba2619ef1153f80d05d67c2ee3a4f
Tracked-On: IMINAN-20365
Signed-off-by: Zaikuo Wang <zaikuo.wang@intel.com>
Reviewed-on: https://icggerrit.ir.intel.com/45944
Reviewed-by: Fang Wang <fang.c.wang@intel.com>
Validation-Android: Wei Tao <wei.tao@intel.com>
Build: IMIN-legacy (SC)
Integrator: Wei Tao <wei.tao@intel.com>
Build: Wei Tao <wei.tao@intel.com>
Reviewed-by: Wei Tao <wei.tao@intel.com>
Maintainer: Wei Tao <wei.tao@intel.com>
---
 .../camera/drivers/media/i2c/ov5693-ecs/ov5693.c   | 31 +++++++++++++++++++---
 .../camera/drivers/media/i2c/ov5693-ecs/ov5693.h   |  2 ++
 2 files changed, 30 insertions(+), 3 deletions(-)

diff --git a/drivers/external_drivers/camera/drivers/media/i2c/ov5693-ecs/ov5693.c b/drivers/external_drivers/camera/drivers/media/i2c/ov5693-ecs/ov5693.c
index b0855ab..455588d 100644
--- a/drivers/external_drivers/camera/drivers/media/i2c/ov5693-ecs/ov5693.c
+++ b/drivers/external_drivers/camera/drivers/media/i2c/ov5693-ecs/ov5693.c
@@ -1387,7 +1387,8 @@ static int __power_up(struct v4l2_subdev *sd)
 		goto fail_power;
 
 	/* according to DS, at least 5ms is needed between DOVDD and PWDN */
-	usleep_range(5000, 6000);
+	/* add this delay time to 10~11ms*/
+	usleep_range(10000, 11000);
 
 	/* gpio ctrl */
 	ret = gpio_ctrl(sd, 1);
@@ -1631,8 +1632,32 @@ static int ov5693_s_mbus_fmt(struct v4l2_subdev *sd,
 	}
 
 	ret = startup(sd);
-	if (ret)
-		dev_err(&client->dev, "ov5693 startup err\n");
+	if (ret) {
+		int i = 0;
+		dev_err(&client->dev, "ov5693 startup err, retry to power up\n");
+		for (i = 0; i < OV5693_POWER_UP_RETRY_NUM; i++) {
+			dev_err(&client->dev,
+				"ov5693 retry to power up %d/%d times, result: ",
+				i+1, OV5693_POWER_UP_RETRY_NUM);
+			power_down(sd);
+			ret = power_up(sd);
+			if (!ret) {
+				mutex_unlock(&dev->input_lock);
+				ov5693_init(sd);
+				mutex_lock(&dev->input_lock);
+			} else {
+				dev_err(&client->dev, "power up failed, continue\n");
+				continue;
+			}
+			ret = startup(sd);
+			if (ret) {
+				dev_err(&client->dev, " startup FAILED!\n");
+			} else {
+				dev_err(&client->dev, " startup SUCCESS!\n");
+				break;
+			}
+		}
+	}
 
 	/*
 	 * After sensor settings are set to HW, sometimes stream is started.
diff --git a/drivers/external_drivers/camera/drivers/media/i2c/ov5693-ecs/ov5693.h b/drivers/external_drivers/camera/drivers/media/i2c/ov5693-ecs/ov5693.h
index 3d88d95..2f14b3d 100755
--- a/drivers/external_drivers/camera/drivers/media/i2c/ov5693-ecs/ov5693.h
+++ b/drivers/external_drivers/camera/drivers/media/i2c/ov5693-ecs/ov5693.h
@@ -36,6 +36,8 @@
 
 #define OV5693_NAME		"ov5693"
 
+#define OV5693_POWER_UP_RETRY_NUM 5
+
 /* Defines for register writes and register array processing */
 #define I2C_MSG_LENGTH		0x2
 #define I2C_RETRY_COUNT		5
-- 
1.9.1

